opm/material/fluidsystems/blackoilpvt/GasPvtThermal.hpp

```diff
@@ -147,12 +147,13 @@ class GasPvtThermal
 
                     // integrate to the heat capacity from the current sampling point to the next
                     // one. this leads to a quadratic polynomial.
-                    Scalar h0 = cvGasColumn[i];
-                    Scalar h1 = cvGasColumn[i + 1];
+                    Scalar c_v0 = cvGasColumn[i];
+                    Scalar c_v1 = cvGasColumn[i + 1];
                     Scalar T0 = temperatureColumn[i];
                     Scalar T1 = temperatureColumn[i + 1];
-                    Scalar m = (h1 - h0)/(T1 - T0);
-                    Scalar deltaU = 0.5*m*(T1*T1 - T0*T0) + h0*(T1 - T0);
+                    Scalar m = (c_v1 - c_v0)/(T1 - T0);
+                    Scalar c = c_v0 - m*T0;
+                    Scalar deltaU = 0.5*m*(T1*T1 - T0*T0) + c*(T1 - T0);
```

Comment:
Two points:
 - I still think the form `Scalar deltaU = 0.5 * (c_v0 + c_v1) * (T1 - T0);` is shorter, more understandable, and less likely to suffer from floating point cancellation.
 - I had no idea this was happening in tons of functions... Seems like a small utility function (with a unit test...) is in order?

Reply:
> I still think the form Scalar deltaU = 0.5 * (c_v0 + c_v1) * (T1 - T0); is shorter, more understandable, and less likely to suffer from floating point cancellation.

I agree. patch pushed. (NB: the result is now a few microseconds faster for thermal runs ;))

> I had no idea this was happening in tons of functions... Seems like a small utility function (with a unit test...) is in order?

A unit test would most likely not have caught the mistake at hand. Rather, it would probably have "petrified" it, so in this case it would have been rather counter productive IMO.

I'm always open for some more unit tests, though, but this is stuff for a different PR.

Reply:
> A unit test would most likely not have caught the mistake at hand. Rather, it would probably have "petrified" it, so in this case it would have been rather counter productive IMO.

Even a unit test on the function `y = x` would have caught this, so I disagree.

> I'm always open for some more unit tests, though, but this is stuff for a different PR.

I can accept that.

Reply:
Unit testing - it as always for the mythical 'other PR'  - or?

Reply:
yeah, same as with documentation ;)

Reply:
> Even a unit test on the function y = x would have caught this,

if the tested interval started at x=0, it would *not* have caught it.

Reply:
I envisioned a utility for "representing the integral of a piecewise linear function as another piecewise linear function", this would have gone wrong as soon as we tested with more than one data interval.